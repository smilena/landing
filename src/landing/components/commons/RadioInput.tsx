import { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import styled from "styled-components";
import { setUserData, UserData } from '../../../redux-state/UserDataSlice';
import { RootState } from '../../../store';

export type RadioInputProps = {
  id: string;
  label: string;
  substep?: RadioInputProps
}

const Components = {
  Wrapper: styled.div`
    display: flex;
    flex-direction: column;
  `,
  OptionWrapper: styled.div`
    display: flex;
    justify-content: space-between;
    margin: 4px;
  `,
  Label: styled.label`
    color: #504437;
    font-size: 18px;
  `,
  Input: styled.input`
    accent-color: #504437;
  `,
}

const mapOptionValues: { [key: string]: boolean } = {
  'true': true,
  'false': false,
}

export const RadioInput = ({id, label, substep}: RadioInputProps) => {
  
  const dispatch = useDispatch();
  const userData:UserData = useSelector((state: RootState) => state.userData);
  const [value, setValue] = useState(userData[id  as keyof UserData ] || false);
  const [subStepValue, setSubStepValue] = useState(false);

  useEffect(() => {
    dispatch(setUserData({id, value}));
    substep && dispatch(setUserData({id: substep?.id, value: subStepValue}));
  }, [value, subStepValue])
  

  const onChangeHandler = (e: React.FormEvent<EventTarget>) => {
    const { value } = e.target as HTMLInputElement;
    setValue(mapOptionValues[value]);
  }

  const onSubStepChangeHandler = () => {
    setSubStepValue(!subStepValue);
  }
  
  return (
    <Components.Wrapper>
      <Components.Label htmlFor={id}>{label}</Components.Label>
      {
        <>
          <Components.OptionWrapper>
            <Components.Label>Si</Components.Label>
            <Components.Input
              type="radio"
              value="true"
              checked={mapOptionValues[`${value}`]}
              onChange={onChangeHandler} />
          </Components.OptionWrapper>
          <Components.OptionWrapper>
            <Components.Label>No</Components.Label>
            <Components.Input
              type="radio"
              value="false"
              checked={!mapOptionValues[`${value}`]}
              onChange={onChangeHandler} />
          </Components.OptionWrapper>
        </>
      }
      {
        substep && value && <Components.OptionWrapper>
          <Components.Label htmlFor={substep.id}>{substep.label}</Components.Label>
          <Components.Input
            type="checkbox"
            value={substep.id}
            checked={subStepValue}
            onChange={onSubStepChangeHandler}
          />
        </Components.OptionWrapper>
      }
    </Components.Wrapper>
  )
}