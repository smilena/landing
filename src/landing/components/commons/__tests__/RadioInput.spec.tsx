import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import createMockStore from 'redux-mock-store';
import { RadioInput } from '../RadioInput';

const mockStore = createMockStore([]);

afterEach(cleanup);

describe('Given RadioInput component', () => {
  test('should render correctly', () => {
    const state = {
      userData: {
        parking: true,
      }
    };
    const store = mockStore(state);
    const { container } = render(
      <Provider store={store}>
        <RadioInput id={'floor'} label={'NÃºmero de piso'}/>
      </Provider>);
    expect(container).toMatchSnapshot();
    expect(screen.getAllByRole('radio')[0]).toBeInTheDocument();
    fireEvent.click(screen.getAllByRole('radio')[0]);
  });
});
