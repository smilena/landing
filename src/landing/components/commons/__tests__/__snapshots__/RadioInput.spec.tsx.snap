// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given RadioInput component should render correctly 1`] = `
<div>
  <div
    class="sc-bczRLJ eaUlQc"
  >
    <label
      class="sc-dkzDqf iguwhg"
      for="floor"
    >
      NÃºmero de piso
    </label>
    <div
      class="sc-gsnTZi kASbhL"
    >
      <label
        class="sc-dkzDqf iguwhg"
      >
        Si
      </label>
      <input
        class="sc-hKMtZM ijbipV"
        type="radio"
        value="true"
      />
    </div>
    <div
      class="sc-gsnTZi kASbhL"
    >
      <label
        class="sc-dkzDqf iguwhg"
      >
        No
      </label>
      <input
        checked=""
        class="sc-hKMtZM ijbipV"
        type="radio"
        value="false"
      />
    </div>
  </div>
</div>
`;
